README.md
## MagicMirror Module: Tabulator


| STATUS: | Version | Date | Maintained? |
|:------- |:------- |:---- |:----------- |
| Working | `1.0.2` | 2018-03-05 | YES |


**What is this module doing?**

`MMM-Tabulator` is just a demonstration module for using the [Tabulator](http://tabulator.info/) *npm* package on your [MagicMirror](https://github.com/MichMich/MagicMirror). 
*Tabulator* is a very nice tool for producing dynamic HTML tables from various sources, like JSON, CSV and XML etc. 
However, although very well documented, the [examples](http://tabulator.info/examples/) are very hard to understand and implement in practice, 
for non-experts. 


**Why I should use Tabulator?**

* Very flexible for manipulating all aspects of table data
* Can sort table columns in several ways and in multiple stages
* Can use graphics in tables
* Can easily enable/disable various table items
* Help reduce all the clutter of HTML generating JS code
* You never have to write DOM code again!
* You don't need to think about converting XML or JSON to HTML table
* Very well used and maintained repository


**Why I should not use Tabulator for my MM:**

* It is totally overkill for showing small and simple tables
* It is a very large library, easily bloating your MMs if used repeatedly
* It is definitely slowing down load times somewhat
* It is dependent on:  *jquery* and *jqeuery-ui*. 


### Example Screenshots

In this repo there are 2 demo examples.

1. The `demo.html` file, that has the JSON embedded and can be run directly from your browser.
2. The `MMM-Tabulator` Module, that is importing the same JSON data from the file `demo.json`.

Screenshot of the `demo.html` output:

![Full](demo1.png)

Screenshot of the `MMM-Tabulator` output:

![Full](demo2.png)


The flight JSON data shown, was provided by [FlightRadar24](https://www.flightradar24.com/). 

---


### Dependencies

This module depend on the following *npm* packages:

* [jquery](https://github.com/jquery/jquery)  - is a fast, small, and feature-rich JavaScript library
* [jquery-ui](https://github.com/jquery/jquery-ui)  - for user interface interactions, effects, widgets, and themes
* [tabulator](https://github.com/olifolkerd/tabulator) - for an overkill use of JSON to HTML table generation


These are also listed in the `package.json` file and should be installed automatically when using *npm*.
However, those may require other packages. 

---


### Installation

Manual Installation:

```bash
cd ~/MagicMirror/modules
git clone https://github.com/E3V3A/MMM-Tabulator.git
cd MMM-Tabulator

# To locally install: jquery, jquery-ui and tabulator, do:
chmod 755 install_deps.sh
./install_deps.sh

# ToDo:
#npm install
```

Easy Installation [**ToDo**]:


```bash
npm install mmm-tabulator
```

### Configuration 

To configure the Tabulator for MM, you need to do the following:

* Add the Module to the global MM *config.js* 
* Edit `node_helper.js` to select the correct JSON file or JSON input file:  

   `const myfile = 'modules/MMM-Tabulator/demo.json';`

* Edit `MMM-Tabulator.js` to:
    - select & configure the table rows you want to show
    - edit the column formatting to be used
    - [TBA] select the filename or URL from where to load the JSON data
    - select the default Tabulator CSS
* Edit `MMM-Tabulator.css` for your own CSS taste

Add the module to the modules array in the `config/config.js` file by adding the following section. 

```javascript
{
    module: 'MMM-Tabulator',
    header: 'Tabulator Demo',
    position: 'top_left',
    config: {
        //fileUrl: "demo.json",     // The local filename or URL to JSON source
        updateInterval: 180,        // [seconds] Read the file every 3 min
        maxItems: 10,               // Max table rows shown
    }
},
...
```

You can change this configuration later when you see that it works.


#### Configuration Options


| Option         | Default          | Description |
|:-------------- |:---------------- |:------------| 
| header         | "Tabulator Demo" | This Module's header text |
| updateInterval | 180              | Module data update in [seconds] |
| maxItems       | 10               | Number of table rows to show |
| fileUrl*       | "demo.json"      | A local filename or URL to a JSON API |

`*` - **not yet implemented - ToDo!**

---

#### `Tabulator` configuration

The main part to configure is found under the `loadTabulator()` function:

```javascript
...
    flightTable.tabulator({
        height:flightTableHeight,           // Set height of table, this enables the Virtual DOM and improves render speed
        //layout:"fitColumns",                // Fit columns to width of table (optional)
        //headerSort:false,                   // Disable header sorter
        resizableColumns:false,             // Disable column resize
        responsiveLayout:true,              // Enable responsive layouts
        placeholder:"No Data Available",    // Display message to user on empty table
        initialSort:[                       // Define the sort order:
            {column:"altitude",     dir:"asc"},     // 1'st
            //{column:"flight",     dir:"desc"},    // 2'nd
            //{column:"bearing",    dir:"asc"},     // 3'rd
        ],
        columns:[
            ...
            {title:"Time", field:"timestamp", headerSort:false, sortable:false, visible:false, responsive:1, formatter:"ep2time"},
...

```


For further details about configuring the many Tabulator options, see:

* http://tabulator.info/
* https://github.com/olifolkerd/tabulator
* http://tabulator.info/docs/3.4?#css
* http://tabulator.info/examples/3.4?#theming

---

#### Bugs and Warnings

:warning: If you provide your own JSON data, other than what is provided in the `demo.json` 
file, then **it has to be in proper JSON format.** For some reason the MM socket seem to not 
accept loosely formatted JSON. There is an extremely simple JSON sanitizer, used (for my own 
purposes), but it doesn't handle any characters outside `[A-Za-z0-9_\-]`. Not sanitizing your
JSON data is a security issue that could lead to code injection! Also check if your source
is using UTF-8 or something else. 

Please refer to the [issue tracker](https://github.com/E3V3A/MMM-Tabulator/issues) for details and updates.


#### Other Modules using *Tabulator*:

* [MMM-FlightsAbove](https://github.com/E3V3A/MMM-FlightsAbove) - A flight radar showing flight & airplane details for planes above a certain location.


#### Contribution

Feel free to post issues and PR's related to this module.  
For all other or general questions, please refer to the [MagicMirror Forum](https://forum.magicmirror.builders/).

#### Credits 

Most grateful thanks to [Ray Wojciechowski](https://github.com/raywo) for sharing his 
module develpment experince and thus helping me solve all my issues. Do check out his 
many great MMM modules in his repository.


#### License 

MIT 


CHANGELOG.md
### MMM-Tabulator Change Log

#### [1.0.2] - 2018-03-05

New Enhancement Release
- Changed updateInterval to use seconds (fixes #3)
- Added maxItems config option (fixes #4)


#### [1.0.0] - 2018-02-21

First working release.

- Recoded all JS (now it works)
- Updated README.md
- Cleanup up dirty JSON file according to PR
- Added screenshot

#### [0.0.1] - 2018-02-04

- Initial version


node_helper.js
/* -----------------------------------------------------------
 * Module:       MMM-Tabulator
 * FileName:     node_helper.js
 * Author:       E:V:A
 * License:      MIT
 * Date:         2018-03-05
 * Version:      1.0.2
 * Description:  A MagicMirror Demo module for using Tabulator
 * Format:       4-space TAB's (no TAB chars), mixed quotes
 *
 * URL1:         https://github.com/E3V3A/MMM-Tabulator
 * URL2:         https://github.com/olifolkerd/tabulator
 * -----------------------------------------------------------
 * Tabulator Requires:
 *      "node_modules/jquery/dist/jquery.min.js",
 *      "node_modules/jquery-ui-dist/jquery-ui.min.js",
 *      "node_modules/jquery-ui-dist/jquery-ui.css",
 *      "node_modules/jquery.tabulator/dist/js/tabulator.js",
 *      "node_modules/jquery.tabulator/dist/css/tabulator.css",
 * -----------------------------------------------------------
 * See:
 *   https://stackoverflow.com/questions/32621988/electron-jquery-is-not-defined?rq=1
 *   https://electronjs.org/docs/faq
*/

'use strict';

const fs = require('fs');

module.exports = NodeHelper.create({

    start: function() {
        this.config = null;
    },

    readData: function() {
        const myfile = 'modules/MMM-Tabulator/demo.json'; // The demo API use improper JSON

        // Consider using readFileSync() and/or using utf8:
        // see: https://stackoverflow.com/questions/10058814/get-data-from-fs-readfile/14078644
        //fs.readFileSync( myfile, 'utf8', (err, data) => {
        //fs.readFile( myfile, 'utf8', (err, data) => {
        fs.readFile( myfile, 'ascii', (err, data) => { // F24 JSON source is probably not UTF-8
            if (err) throw err;
            //console.log(data);
            let cleanData = jZen(data);
            if (isJSON(cleanData) ) {
                this.sendSocketNotification("NEW_DATA", cleanData);
            } else {
                // So WTF is it?
                console.log("JSON: false");
                console.log("isAO(dirty): " + isAO(data));
                console.log("isAO(clean): " + isAO(cleanData));
                console.log("Data:\n" + data);
            }
        });
    },

    socketNotificationReceived: function (notification, payload) {
        if (notification === "REQUEST_DATA") {
            this.config = payload;
            this.readData();
            setInterval(() => {
                this.readData();
            }, this.config.updateInterval * 1000);
        }
    }

});

// To check if something is JSON
function isJSON(str) {
    try { return (JSON.parse(str) && !!str); }
    catch (e) { return false; }
}

// To check if something is an Array or Object (parsed JSON)
function isAO(val) {
    return val instanceof Array || val instanceof Object ? true : false;
}

// --------------------------------------------------------------------------
// What:  A dirt simple JSON cleanup function that also compactifies the data
// NOTE:  - Only use on flat and trustworthy ASCII JSON data!
//        - Cannot handle any characters outside [A-Za-z0-9_\-]. (e.g. UTF-8)
//        - Using remote data without further sanitation is a security risk!
// --------------------------------------------------------------------------
const re1 = /([A-Za-z0-9_\-]+):/gm;  // use const to make sure it is compiled
function jZen(data) {
    //let re1 = /([A-Za-z0-9_\-]+):/gm; // Find all ASCII words $1 before an ":"
    let str = "";
    str = data.replace(/\s/gm, '');     // Remove all white-space
    str = str.replace(/\'/gm, '\"');    // Replace all ' with "
    str = str.replace(re1, '\"$1\":');  // Replace $1: with "$1":
    //console.log("Dirty JSON is:\n" + data.toString() );
    //console.log("Clean JSON is:\n" + str);
    return str;
}


MMM-Tabulator.js
/* -----------------------------------------------------------
 * Module:       MMM-Tabulator
 * FileName:     MMM-Tabulator.js
 * Author:       E:V:A
 * License:      MIT
 * Date:         2018-03-05
 * Version:      1.0.2
 * Description:  A MagicMirror Demo module for using Tabulator
 * Format:       4-space TAB's (no TAB chars), mixed quotes
 *
 * URL1:         https://github.com/E3V3A/MMM-Tabulator
 * URL2:         http://tabulator.info/
 * URL3:         https://github.com/olifolkerd/tabulator
 * -----------------------------------------------------------
 * About Tabulator CSS & Themes:
 *   http://tabulator.info/docs/3.4?#css
 *   http://tabulator.info/examples/3.4?#theming
 * -----------------------------------------------------------
 */

//-----------------------------------------------------
// Tabulator Requires:
// "node_modules/jquery/dist/jquery.min.js",
// "node_modules/jquery-ui-dist/jquery-ui.min.js",
// "node_modules/jquery-ui-dist/jquery-ui.css",
// "node_modules/jquery.tabulator/dist/js/tabulator.js",
// "node_modules/jquery.tabulator/dist/css/tabulator.css",
//-----------------------------------------------------

'use strict';

Module.register('MMM-Tabulator',{

    defaults: {
        header: "Some Flights", // The module header text, if any
        maxItems: 10,           // MAX number of planes (table rows) to show
        updateInterval: 180     // [sec] Read the file every 3 min
        //fileUrl: "file:///home/pi/MagicMirror/modules/MMM-backlog/demo.json"
    },

    requiresVersion: "2.1.0",

    start: function() {
        this.loaded = false;
        this.sendSocketNotification("REQUEST_DATA", this.config);
    },

    getDom: function() {
        let w = document.createElement("div");  // Let w be the "wrapper"
        w.id = "flighttable";                   // The id used by Tabulator

        if (!this.loaded) {
            w.innerHTML = "Loading...";
            w.className = "dimmed light small";
            return w;
        }
        if (!this.data) {
            w.innerHTML = "No data!";
            w.className = "dimmed light small";
            return w;
        }
        w.innerHTML = "Waiting for Tabulator...";
        return w;
    },

    getScripts: function() {
        return [
            this.file('node_modules/jquery/dist/jquery.min.js'),
            this.file('node_modules/jquery-ui-dist/jquery-ui.min.js'),
            this.file('node_modules/jquery.tabulator/dist/js/tabulator.js')
        ]
    },

    getStyles: function() {
        return [
            this.file('node_modules/jquery-ui-dist/jquery-ui.css'),
            //this.file('node_modules/jquery.tabulator/dist/css/tabulator.css'),                // Standard Theme
            this.file('node_modules/jquery.tabulator/dist/css/tabulator_midnight.min.css'),     // Midnight Theme
            //this.file('node_modules/jquery.tabulator/tabulator_simple.min.css'),              // Simple Theme
            //this.file('node_modules/jquery.tabulator/bootstrap/tabulator_bootstrap.min.css'), // Bootstrap Theme
            "MMM-Tabulator.css"
        ];
    },

    getTranslations: function() { return false; }, // Nothing to translate

    // This come from the MM CORE or from other modules
    notificationReceived: function (notification, payload, sender) {
        if (notification === "DOM_OBJECTS_CREATED") {
            // The div with id "flighttable" now exists, so we can load Tabulate.
            this.loadTabulate();
        }
    },

    // This comes from YOUR module, usually "node_helper.js"
    socketNotificationReceived: function(notification, payload) {
        switch (notification) {
            case "NEW_DATA":
                //console.log("MMM-Tabulator: NEW_DATA received!");
                this.loaded = true;
                this.setTableData(payload);
                break;
        }
    },

    //===================================================================================
    //  From here:  Tabulator specific code
    //===================================================================================
    loadTabulate: function() {

        // see: http://tabulator.info/docs/3.3#mutators
        /*Tabulator.extendExtension("mutator", "mutators", {
            ft2met:function(value, data, type, mutatorParams){
                return (value * 0.3048).toFixed(0);
            },
        });*/

        let self = this;

        Tabulator.extendExtension("format", "formatters", {
            ft2mt:function(cell, formatterParams){              // Feet to Meters
                return  (0.3048*cell.getValue()).toFixed(0);
            },
            kn2km:function(cell, formatterParams){              // Knots to Kilometers
                return  (1.852*cell.getValue()).toFixed(0);
            },
            ep2time:function(cell, formatterParams){            // POSIX epoch to hh:mm:ss
                let date = new Date(cell.getValue());
                // We use "en-GB" only to get the correct formatting for a 24 hr clock, not your TZ.
                return date.toLocaleString('en-GB', { hour:'numeric', minute:'numeric', second:'numeric', hour12:false } );
            },
            // placeholder: deg2dir
        });

        let flightTable = $("#flighttable");
        var flightTableHeight = ( this.config.maxItems * 33 + 33 );   // @12px font-size we have [~33 px/row]

        flightTable.tabulator({
            height:flightTableHeight,           // [px] Set MAX height of table, this enables the Virtual DOM and improves render speed
            height:205,                         // Set height of table, this enables the Virtual DOM and improves render speed
            //layout:"fitColumns",                // Fit columns to width of table (optional)
            //headerSort:false,                   // Disable header sorter
            resizableColumns:false,             // Disable column resize
            responsiveLayout:true,              // Enable responsive layouts
            placeholder:"Waiting for data...",  // Display message to user on empty table
            initialSort:[                       // Define the sort order:
                {column:"altitude",     dir:"asc"},     // 1'st
                //{column:"flight",     dir:"desc"},    // 2'nd
                //{column:"bearing",    dir:"asc"},     // 3'rd
            ],
            columns:[
                {title:"Flight",        field:"flight",         headerSort:false, sortable:false, responsive:0, align:"left"}, // , width:250},
                {title:"CallSig",       field:"callsign",       headerSort:false, sortable:false, visible:true, responsive:3},
                {title:"To",            field:"destination",    headerSort:false, sortable:false, responsive:0},
                {title:"From",          field:"origin",         headerSort:false, sortable:false, responsive:0},
                {title:"Speed",         field:"speed",          headerSort:false, sortable:false, responsive:2, formatter:"kn2km"}, // [km/h]
                {title:"Bearing",       field:"bearing",        headerSort:false, sortable:false, responsive:1},
                {title:"Alt [m]",       field:"altitude",       headerSort:false, sortable:false, responsive:0, formatter:"ft2mt", align:"right", sorter:"number"},
                //{title:"Alt [m]",       field:"altitude",       sortable:true,  responsive:0, align:"right", sorter:"number", mutateType:"data", mutator:ft2met"},
                // Additional items:
                {title:"F24id",         field:"id",             headerSort:false, sortable:false, visible:false},
                {title:"RegID",         field:"registration",   headerSort:false, sortable:false, visible:false},
                {title:"Model",         field:"model",          headerSort:false, sortable:false, visible:true,  responsive:1},
                {title:"ModeS",         field:"modeSCode",      headerSort:false, sortable:false, visible:false},
                {title:"Radar",         field:"radar",          headerSort:false, sortable:false, visible:false},
                {title:"Lat",           field:"latitude",       headerSort:false, sortable:false, visible:false},
                {title:"Lon",           field:"longitude",      headerSort:false, sortable:false, visible:false},

                {title:"Time",          field:"timestamp",      headerSort:false, sortable:false, visible:false, responsive:1, formatter:"ep2time"},
                {title:"RoC [ft/m]",    field:"climb",          headerSort:false, sortable:false, visible:false},
                {title:"Squawk",        field:"squawk",         headerSort:false, sortable:false, visible:true, responsive:1}, // formatter:"sqCheck"},
                {title:"isGND",         field:"ground",         headerSort:false, sortable:false, visible:false},
                {title:"isGlider",      field:"glider",         headerSort:false, sortable:false, visible:false},
            ],
        });

        $(window).resize(function () {
            flightTable.tabulator("redraw");
        });
    },

    setTableData: function(data) {
        $("#flighttable").tabulator("setData", data);
    }
    //===================================================================================

});
